Hide default assessment set warning to students

ZacWarham opened the pull request:
Attempting a method of hiding the default message that is shown to students for missing assessment sets

Closes #6486

nwalters512 commented:
Before you go too far down this road, we added the `implicit` column precisely to make this possible and easy:

https://github.com/PrairieLearn/PrairieLearn/blob/26613d46b786544e5ad34648d15b13987bab1ed4/database/tables/assessment_sets.pg#L7

Note that using the string `"null"` as a sentinel value is a poor choice. What happens when someone uses the string `"null"` as the actual description in their JSON file?

ZacWarham commented:
> What happens when someone uses the string `"null"` as the actual description in their JSON file?

As far as I can tell it creates the auto-generated assessment set called "null" and works as intended which confused me knowing it was a dumb choice because either I found a bug or more likely I am missing something obvious. That is why I put it here for the CLI checks to try and work out what I might be missing (since the tests do not work in Docker).

![assessment list](https://github.com/user-attachments/assets/e0c24c39-2d39-4c15-b592-afcb45d029bb)

![assessment JSON](https://github.com/user-attachments/assets/220f2c2c-e31c-4da8-94ee-fddd89d75398)

If you have any insight there I would appreciate it. In the meantime I will take a look at the column you recommended as the better alternative.

nwalters512 commented:
The text "Auto-generated from..." only appears if an assessment set is _not_ explicitly listed in `infoCourse.json`. The situation I proposed is someone putting the following in `infoCourse.json`:

```json
{
  "assessmentSets": [{
    "name": "null",
    "heading": "null",
    "color": "gray1"
  }]
}
```

Remember that the string `"null"` is not the same as the null value `null`.

ZacWarham commented:
Ah gotcha. I knew it was not the same which is why I was confused. Turns out I checked
- "null" in `infoAssessment.json`
- "null" in `infoAssessment.json` and `infoCourse.json`

But forgot to check

- "null" in `infoCourse.json`

That then breaks as expected and was what I was missing

![broken view](https://github.com/user-attachments/assets/c8be2e8c-38ab-497e-b352-96ab45fccfc4)

ZacWarham commented:
I think this is working now using the `implicit` value from Nathan's suggestion

> https://github.com/PrairieLearn/PrairieLearn/blob/26613d46b786544e5ad34648d15b13987bab1ed4/database/tables/assessment_sets.pg#L7
> 

However it [failed the tests](https://github.com/PrairieLearn/PrairieLearn/actions/runs/13358654907/job/37304920062). I don't understand the tests well enough to know if this is because it is 1) checking the student view (where `Auto-generated from ...` does not exist anymore) or 2) if it is checking the instructor view and I broke it in another way.

I modified the value of the test in https://github.com/PrairieLearn/PrairieLearn/pull/11381/commits/20b8d53621f754f398ba969c912551366311497c and now it passes. Assuming it failed from my former suggestion above, this should be okay? I will keep looking at the tests to try to understand it better but if you have the time @nwalters512, I would appreciate any further clarification on this.

ZacWarham commented:
I have opened this for review to get a second pair of eyes on the issues with the testing mentioned in https://github.com/PrairieLearn/PrairieLearn/pull/11381#issuecomment-2661656881

nwalters512 commented:
> However it [failed the tests](https://github.com/PrairieLearn/PrairieLearn/actions/runs/13358654907/job/37304920062). I don't understand the tests well enough to know if this is because it is 1) checking the student view (where Auto-generated from ... does not exist anymore) or 2) if it is checking the instructor view and I broke it in another way.

Your changes to `apps/prairielearn/src/sync/fromDisk/assessmentSets.ts` very very explicitly no longer write the text `(Auto-generated from use ...` into the database, so it's expected that the tests (which assert things about the contents of the database) would have to be updated.

I'll give this PR a look now. Thanks for your patience.

ZacWarham commented:
> into the database, so it's expected that the tests (which assert things about the contents of the database) would have to be updated.

Thanks for the clarification here. I thought it was checking the UI view, not the database itself so that makes sense.

nwalters512 commented on apps/prairielearn/src/pages/instructorAssessments/instructorAssessments.html.ts:
```diff
@@ -117,7 +117,12 @@ export function InstructorAssessments({
                           ? html`
                               <tr>
-                                <th colspan="7" scope="row">${row.assessment_group_heading}</th>
+                                <th colspan="7" scope="row">
+                                  ${!row.implicit
+                                    ? row.assessment_group_heading
+                                    : row.abbreviation +
+                                      ' (Auto-generated from use in an assessment; add this assessment set to your infoCourse.json file to customize)'}
+                                </th>
```
I don't believe this does the correct thing when a course instance is configured to group assessments by assessment module.

To fix this, you might consider getting rid of `assessment_group_heading` from the query and moving the decision about whether to render the set heading or module heading into TypeScript.

ZacWarham replied:
Ah, I have not used a course with Assessment Modules so I did not consider this. I will set one up in my local instance and take a look.

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminSets/instructorCourseAdminSets.html.ts:
```diff
@@ -54,7 +54,12 @@ export function InstructorCourseAdminSets({
                     <td class="align-middle">${assessment_set.name}</td>
-                    <td class="align-middle">${assessment_set.heading}</td>
+                    <td class="align-middle">
+                      ${!assessment_set.implicit
+                        ? assessment_set.heading
+                        : assessment_set.abbreviation +
+                          ' (Auto-generated from use in an assessment; add this assessment set to your infoCourse.json file to customize)'}
```
I'd recommend adding an `AssessmentSetHeading` component in the `components/` directory that can do this kind of thing. This code is repeated verbatim in two files; it'd be nice to reuse it. In the future, we'll probably want to include a link to an in-browser editor for assessment sets which will be even more complex and thus we'd get more value from sharing it.

As part of that, consider wrapping the explanatory text in `<span class="text-muted"></span>`. IMO this helps call out that this isn't actually part of the heading.

Without my suggested change:

<img width="1100" alt="Screenshot 2025-02-28 at 15 08 10" src="https://github.com/user-attachments/assets/64dc893a-466a-47b8-b104-e640b846262c" />

With my suggested change:

<img width="1101" alt="Screenshot 2025-02-28 at 15 10 12" src="https://github.com/user-attachments/assets/cba51b13-a42d-451d-bbc3-bd6cc4b1cf54" />

nwalters512 replied:
Sorry for not being sufficiently clear here. I meant a literal component called `AssessmentSetHeading` that would be passed an `assessment_sets` row from the database and would decide how to render it, regardless of whether it's implicit or not.

ZacWarham replied:
@nwalters512 I want to make sure I am understanding you correctly so I have just made a small change here which I think is in the right direction? - https://github.com/PrairieLearn/PrairieLearn/pull/11381/commits/9ac51341fe0a8d252828df1a312e359a0eb7e688 - This is my first time working with TypeScript components so I may not fully understand your request sorry. 

I am also uncertain what you mean by 

> regardless of whether it's implicit or not.

when you said previously

> we added the `implicit` column precisely to make this possible

I assume you mean to remove the use of implicit inside of the `return html ... ` and then return one `<tr>` if implicit and another `<tr>` if not implicit? And then also change it depending if it is sorted by module or not.

nwalters512 replied:
I guess I was still insufficiently clear.

```ts
                                  ${!row.implicit
                                    ? row.assessment_group_heading
                                    : row.abbreviation +
                                      ' (Auto-generated from use in an assessment; add this assessment set to your infoCourse.json file to customize)'}
```

Before your recent changes, this exact code was duplicated in two places. My request was that you move this code (the piece that determines what is rendered as a set's "heading") to a component. That component would be called `AssessmentSetHeading` and would contain more or less exactly the above logic. You can then use this on both the instructor assessment page and on the assessment sets page.

nwalters512 reviewed (commented):
This looks more or less correct, minus the handling of assessment modules.

Once this lands, would you be willing/able to contribute similar changes to assessment modules and question tags/topics? They have very similar behavior (and current deficiencies).

ZacWarham commented:
I have moved the explanatory text into a span inside of the `AssessmentSetHeading.html.ts` component with the function `MissingDefinition` as per https://github.com/PrairieLearn/PrairieLearn/pull/11381#discussion_r1976108288 - not sure if you have any standards for naming conventions so please let me know if you want that changed.

In fixing the issue highlighted in https://github.com/PrairieLearn/PrairieLearn/pull/11381#discussion_r1976106707 I have also expanded the scope of this PR to now cover modules.

ZacWarham commented:
Thanks for your patience with this @nwalters512 - I have updated this now after our discussion today. Please let me know if you have any further suggestions or fixes.

I also was wondering if it should handle the 'Default Module' differently? Given that an assessment module is not mandatory, when sorting by modules the 'Default Module' is applied  to an assessment if one does not exist. Should the message say something like `A module has not been defined for this assessment, please define one in the assessmentInfo.json` for these cases

![default module](https://github.com/user-attachments/assets/aa7fedbb-af2c-49e8-a576-c7d895003061)

nwalters512 commented:
I think it's worth special casing the default module, though I'd rather just omit the "Auto-generated..." message entirely. It's perfectly valid to specify `"module": "Default"` in an `infoAssessment.json`.

Specifically: don't show that extra message if `name === 'Default' && implicit`.

I'll review the other changes here now.

ZacWarham commented on apps/prairielearn/src/pages/instructorAssessments/instructorAssessments.html.ts:
```diff
@@ -117,7 +120,13 @@ export function InstructorAssessments({
+                                  ${assessmentsGroupBy === 'Set'
+                                    ? AssessmentSetHeading({ assessment_set: row.assessment_set })
+                                    : AssessmentModuleHeading({
+                                        assessment_module: row.assessment_set,
```
Dumb luck (or not enough checking things) by the looks of it!

ZacWarham replied:
Fixed in dbd98d8

ZacWarham commented:
This should be most of the way there now after the previous round of feedback. Still need to work on making sure the default module works for a user defined `Default`, a non-defined module `Default` and for the already existing entries that include the string in the database

```
if (!assessment_module.implicit || assessment_module.name === 'Default') {
    return assessment_module.heading;
  }
```

This returning `Default` when a user has defined `Default` inside of assessmentInfo.json instead of showing the error or showing the heading of the `Default` module

My implementation (defined)

![image](https://github.com/user-attachments/assets/9d85d3dd-6362-45b0-b458-abee9c59c378)

My implementation (not defined)

![image](https://github.com/user-attachments/assets/677ecfea-ee0b-4df8-87cf-6fb00e770a0d)


Current implementation (defined)

![image](https://github.com/user-attachments/assets/f59b7c28-5033-4d48-abaf-18d5ef2fca96)

Current implementation (not defined)

![image](https://github.com/user-attachments/assets/da726140-7c67-42c3-b97a-ca06f664765b)

ZacWarham commented:
Using `name === 'Default' && implicit` caused the issues with the name being shown as the heading for `Default` Modules. Switching to the following fixes this
```
assessment_module.heading === 'Default module' && assessment_module.implicit
```

nwalters512 commented on apps/prairielearn/src/pages/instructorAssessments/instructorAssessments.html.ts:
```diff
@@ -27,10 +29,11 @@ export const AssessmentStatsRowSchema = AssessmentSchema.extend({
 
 export const AssessmentRowSchema = AssessmentStatsRowSchema.merge(
-  AssessmentSetSchema.pick({ abbreviation: true, name: true, color: true }),
+  AssessmentSetSchema.pick({ abbreviation: true, name: true, color: true, implicit: true }),
```
@ZacWarham please address this comment.

ZacWarham replied:
Fixed in 589e85a

nwalters512 commented:
@ZacWarham there are still unaddressed comments. For both my sake and yours, it's helpful to comment "Resolved in [commit hash]" once you've resolved something; that'll let me verify your changes, and help you keep track of what you still have to address. You can see an example of this kind of exchange here: https://github.com/PrairieLearn/PrairieLearn/pull/11518#discussion_r1987755120

ZacWarham commented on apps/prairielearn/src/components/AssessmentSetHeading.html.ts:
```diff
@@ -0,0 +1,15 @@
+  return html`
+    ${!assessment_set.implicit
+      ? assessment_set.heading
+      : html`${assessment_set.abbreviation}
```
Fixed in dbd98d8

ZacWarham commented on apps/prairielearn/src/components/AssessmentModuleHeading.html.ts:
```diff
@@ -0,0 +1,19 @@
+  return html`
+    ${!assessment_module.implicit
+      ? assessment_module.heading
+      : html`${assessment_module.heading}
```
Fixed in dbd98d8

ZacWarham commented:
Sorry @nwalters512 - I was using the 🚀 to indicate to myself what I had done to avoid  sending an email every time a comment is added. I also misunderstood your comment as "remove my change" not "remove this section of code"

I have fixed that in 589e85a and gone and added specific commits to each comment.

Also;
> Specifically: don't show that extra message if name === 'Default' && implicit.

Fixed in 9a6e84c as per https://github.com/PrairieLearn/PrairieLearn/pull/11381#issuecomment-2715950122

nwalters512 commented on apps/prairielearn/src/pages/instructorAssessments/instructorAssessments.html.ts:
```diff
@@ -118,13 +120,21 @@ export function InstructorAssessments({
                         <tr id="row-${row.id}">
                           <td class="align-middle" style="width: 1%">
-                            <span class="badge color-${row.color}">${row.label}</span>
+                            <span class="badge color-${row.assessment_set.color}"
+                              >${row.label}</span
+                            >
```
Formatting nit:

```suggestion
                            <span class="badge color-${row.assessment_set.color}">
                              ${row.label}
                            </span>
```

nwalters512 commented on apps/prairielearn/src/components/AssessmentModuleHeading.html.ts:
```diff
@@ -0,0 +1,24 @@
+  assessment_module: AssessmentModule;
+}) {
+  if (
+    !assessment_module.implicit ||
+    (assessment_module.heading === 'Default module' && assessment_module.implicit)
```
Nit, this is equivalent:

```suggestion
    !assessment_module.implicit || assessment_module.heading === 'Default '
```

ZacWarham replied:
This comes from here (which may want to be made a capital `m` to fit the theme of the example course) 
https://github.com/PrairieLearn/PrairieLearn/blob/4b20b51af34b173469b0be56328d6dc0b0a101b5/apps/prairielearn/src/sync/fromDisk/assessmentModules.ts#L52

nwalters512 replied:
Say less! That was just a bad edit on my part, I dropped the trailing "module" when collapsing lines. I edited my suggestion to correct that; the overall idea of dropping the explicit check for `assessment_module.implicit` in the second branch of the `||` was the important piece of my suggestion.

nwalters512 reviewed (approved):
I tested this locally and it's working well. Thanks for working through the review process with me. Two final nits; let me know when you're ready for me to merge this.

ZacWarham commented:
> Thanks for working through the review process with me. Two final nits; let me know when you're ready for me to merge this.

Not at a problem! I appreciate you taking the time to teach me through it. This one is ready to merge and I drafted up a components PR in #11595 for topics/tag if you have any specific requirements for them. @nwalters512
